src/lem/eval.rs

```diff
@@ -1029,22 +977,30 @@ fn reduce(cprocs: &[(&Symbol, usize)]) -> Func {
                         }
                         match symbol head {
                             "lambda" => {
-                                let (args, body) = car_cdr(rest);
-                                let (arg, cdr_args) = extract_arg(args);
-
-                                match arg.tag {
-                                    Expr::Sym => {
-                                        match cdr_args.tag {
-                                            Expr::Nil => {
-                                                let function: Expr::Fun = cons3(arg, body, env);
-                                                return (function, env, cont, apply)
+                                let (vars, rest) = car_cdr(rest);
+                                let rest_nil = eq_tag(rest, nil);
+                                if rest_nil {
+                                    return (expr, env, err, errctrl)
+                                }
+                                let (body, end) = car_cdr(rest);
+                                let end_nil = eq_tag(end, nil);
+                                if !end_nil {
+                                    return (expr, env, err, errctrl)
+                                }
+                                match vars.tag {
+                                    Expr::Cons => {
+                                        let (var, _rest_vars) = decons2(vars);
+                                        match var.tag {
+                                            Expr::Sym => {
+                                                let fun: Expr::Fun = cons4(vars, body, env, foo);
+                                                return (fun, env, cont, apply)
```

Comment:
I think this check is not necessary and we could go straight to
```rust
let fun: Expr::Fun = cons4(vars, body, env, foo);
return (fun, env, cont, apply)
```
It will save a few constraints by deferring the error, which *will* happen at some point if some variable is not a symbol.
I'm following the rule of thumb "allow the execution of correct Lurk programs to be proved faster by the cost of delaying the failure of incorrect Lurk programs"

Reply:
It will save some constraints, but it is too minimal for it to matter. The way it is now more closely follows current Lurk semantics. For instance, in both current Lurk and this branch, `(lambda (x 0) x)` does *not* fail, but  `(lambda (0) 0)` and `((lambda (x 0) x) 0)` fail on both. If I remove this line (and the similar one in apply_cont), these last two expressions will not fail

Reply:
I don't see the practical benefit of `(lambda (0 x) nil)` failing earlier if `(lambda (x 0) nil)` doesn't, especially since this asymmetry has extra costs

Reply:
I agree with @gabriel-barrett for two reasons, based on my understanding from offline discussion:
- We should studiously avoid *gratuitous* changes to Lurk Alpha reduction behavior. We *can* make changes, but those need to be strongly justified as carrying their weight in terms of optimization or other benefit. Every change we make to fundamental reduction behavior is inherently negative — and the more time we spend revisiting basic semantic decisions, the less we have available for other purposes.
- Apart from the general principle, the actual details of this case support Lurk Alpha's original decision. The Lurk reduction step should never create 'bad data' from 'good data'. In this case, a lambda list containing an illegal value is perfectly good Lurk data. It happens to be good Lurk data specifying a 'bad program' because it contains invalid syntax. If the Lurk reduction then creates a malformed function object (`Fun`) that should never exist, then the Lurk reduction is violating a language invariant.

Reply:
* I don't think what I am proposing is a gratuitous change because
* This optimization already violates your second bullet point

Let me elaborate. With this optimization, `(lambda (x 0) nil)` reduces to the multi-argument `<FUNCTION (x 0) nil>`, containing a `0` as one of its variables. This is arguably "bad data".

Lurk Alpha had a compelling reason to make `(lambda (0 x) nil)` error instantly. Its functions were always unary (even if a dummy variable was needed for this) and the symbol verification was performed at lambda -> function reduction time (and nowhere else!).

This PR implements the variable tag check at binding time, here:
https://github.com/lurk-lab/lurk-rs/blob/f7805c62ac2da0077bffa73fa76bff946ca4ed45/src/lem/eval.rs#L1301-L1312

Meaning that the proposed model checks the tag of the head variable twice.

Reply:
I see.

Reply:
We can only fix this if we do argument checking, which will increase the number of iterations. The check costs 50 constraints and 20 auxiliaries, which is not too much. Here's what I propose: keep this check so that it is behaviourally equal to Alpha, and doesn't break any of the tests (if I change, it will break one of the tests). It also makes Lurk less weird.

Once we get a smaller circuit, these constraints might become relevant, so we can do a review of the whole circuit and decide whether or not to reduce checks to get less constraints. Is that fine?

Reply:
Now that I've expressed my points and we're all on the same boat, aware of what's happening, I'm cool with any path you guys pick.

Btw, I don't think that's exactly broken. It depends on how you look at it.

If we want to help the users with early errors of this kind, this could be raised at parsing time (or interning time). And we could check *all* variables without any extra cost on the circuit. This check could also be performed for `let` and `letrec`, making the following error before evaluation even starts:
```lisp
(let ((a 1) (0 0)) nil)
```

Reply:
Yeah, I think this is a good idea. We should check this at parsing time

Reply:
I actually disagree. 'read time' needs to not impose any syntax because that's not how Lisp works. I think it's fine to have a provable compilation step, after which the result is assumed to have been checked, though.